Project_For_Binod/SignUpViewModel.swift

```diff
@@ -17,8 +17,44 @@ enum SignUpEvent {
   case statusLabelTapped
 }
 
+protocol SignUpStateProtocol {
+  var emailText: String { get }
+  var fullnameText: String { get }
+  var passwordText: String { get }
+  var colorText: String { get }
+}
+
+extension SignUpStateProtocol {
+  private var isEmailValid: Bool {
+    return emailText.contains("@") && emailText.contains(".")
+  }
+  
+  var isFullNameValid: Bool {
+    return fullnameText.contains(" ")
+  }
+  
+  var isPasswordValid: Bool {
+    return passwordText.count > 5
+  }
+  
+  var isColorValid: Bool {
+    return !colorText.isEmpty
+  }
+  
+  var isSignUpValid: Bool {
+    return isEmailValid && isFullNameValid && isPasswordValid && isColorValid
+  }
+}
+
 struct SignUpViewModel {
   
+  struct UIState: SignUpStateProtocol {
+    var emailText: String
+    var fullnameText: String
+    var passwordText: String
+    var colorText: String
+  }
+  
   // MARK: Properties
   private let userSessionRepository: UserSessionRepository
```

Comment:
should make this an interface which is specific to what the signup screen needs rather than a general user session repository, since:

*  the user is not logged in
* there should be no session-related functions visible to this class
* the class should only have signin-related functionality visible to it